### ISSUE TITLE
Default security level prevents use of *client* certificate.

### ISSUE THREAD

**Issue by dwmw2** [openssl/openssl#11702]
In https://gitlab.com/openconnect/openconnect/-/issues/132 a user reports that they are unable to use the SSL client certificate which they need to connect to their VPN, because OpenSSL's default security level prevents it.

Ultimately, that is a decision for the *server* not the client to make. The security level configured in OpenSSL locally should affect which *server* certificates we're willing to negotiate, sure. And it should affect our negotiation of ciphersuites, etc.

But why should it prevent us from presenting a client certificate on the wire for the *server* to decide whether it trusts? We're allowed to present a client certificate for which we don't trust the issuing CA; why shouldn't we be allowed to present a client certificate for which we wouldn't accept the issuing hash algorithm?

We've tried working around it by setting the security level to zero just for the duration of the call to `SSL_CTX_use_certificate()` but it still fails in `SSL_connect()`.

Is there a way to work around this?

**Comment by kroeckx** [comment]
I'm not sure why you think there should be a difference in
behaviour between the client and the server. If the client says
they want to reach a certain security level, why would would it be
ok to use an old certificate that doesn't reach that level?

**Comment by dwmw2** [comment]
Because that's a decision for the *server*. Because in the context of a Linux distribution, setting a default security shouldn't necessarily mean "You cannot now authenticate to VPN servers which require the certificate which you are explicitly trying to use".

In particular, I'd love to have a way to use the certificate which the user has explicitly asked me to use, but *without* just setting the security level to zero even for the things which *are* the client's decision, like ciphersuite negotiation and which server certificates it'll accept

**Comment by dlenski** [comment]
> If the client says
> they want to reach a certain security level, why would would it be
> ok to use an old certificate that doesn't reach that level?

But isn't the main point of saying “I want to reach a certain security level” that OpenSSL will check whether the _other_ participant(s) in the connection do something to reduce the security below a certain level? (e.g. client offers 3 high-security ciphersuites, and one lower-security ciphersuite, server picks the lower-security option)

**Comment by kaduk** [comment]
I'm still not seeing a strong argument for why "that's a decision for the server" applies to the described situation.  Yes, it's ultimately the server's decision to decide whether or not to accept a given certificate, but the client's setting of security level seems like it should apply to all algorithms used on the connection (in the "weakest link" model), which would include the client's certificate.
There are many ways in which someone can configure internally inconsistent configuration, with "client certificate does not match security level" just one of them; we can report the inconsistency as an error when it is detected and leave the user to decide which side of the inconsistency to resolve.

**Comment by dwmw2** [comment]
Sure, but the client currently has no way to indicate that it wants to stick to the specified security level *except* for using the explicitly requested certificate. My only option is just to reduce the security level completely.

**Comment by kaduk** [comment]
Would you say it more accurately reflects the reality of that connection that the security level is reduced?  Note that (IIRC) the security level is just a convenient shorthand for applying several common levels of configuration at once, but the specifics could still be applied individually even at a more tolerant overall security level.

**Comment by dwmw2** [comment]
Not really. If an OS distribution sets a higher security level it just places a burden on the user, and complexity in the user interface of the software, to make software use the certificate which was issued for authentication and which is entirely outside the control of the user themselves and of the software on the system in question.

**Comment by kaduk** [comment]
I'm not really sure that I want to consider the "security level set by the OS vendor" caveat for this distinction, and would rather have an answer to "if we had to pick a number to describe the security level of a connection meeting these criteria, what number would it be?".  Whether OS vendors are right to impose this kind of constraint seems something of a separable question.

**Comment by t8m** [comment]
If an OS vendor sets a higher security level, they usually "mean it". And changing the code to ignore the level for the client certificate on the client side would be certainly undesirable for the OS vendor (I am talking here with my Fedora/RHEL hat on). Or in general - the level should not just affect what we consume, but also what we produce. The user still has the option to lower the level.

**Comment by dwmw2** [comment]
@t8m If we were having that discussion in the Fedora context I would argue very strongly against it. The security level of the distro overall is one thing, but it must not *prevent* users from being able to connect to their corporate VPN, using the certificate that their employer issues to them.

As for also affecting what we produce, I just don't agree. I don't expect to do CRL checking on the client certificate, I don't expect the client to refuse to provide the client certificate unless it's issued by a CA which is trusted on the *client* system either. Neither do I expect clients to *enforce* my password strength. All those are up to the server.

But even if I concede that point, and if the user needs to make an *exception* in order to use that certificate, that's tolerable. But they shouldn't need to reduce the security level of the *whole* system just to do so. In fact they shouldn't even need to reduce the overall security level *even* of the VPN client process. There should be a way for it to honour the security level in all other respects *except* for the certificate it's explicitly been asked to use. That's all I'm really looking for.

Temporarily setting the security level to zero around the `SSL_CTX_use_certificate()` call *almost* worked, but fails in `SSL_connect()` later. Can we bypass that later check perhaps, on the basis that the certificate will have been 'vetted' when it was first added? Or some slightly more involved way to mark it as an explicit exception?

**Comment by dwmw2** [comment]
>  the level should not just affect what we consume, but also what we produce.

Put another way: we are *consuming* the client certificate which was issued to us. We aren't really *producing* it. We're just demonstrating that we are in possession of it. By refusing to use it, we are not being liberal in what we accept.

**Comment by richsalz** [comment]
I think saying that the application is consuming the certificate is using rather tortured logic to try to make a point.

For the use-case you mention, if the enterprises issues a certificate that is below the security posture of the employee's machine, then they should tell the employee how to change things. The client machine security posture must be part of the operating environment considered by the enterprise. In a BYOD world, employees might want stricter security, even if they have to "downgrade" it assert their company identity.

I am not a fan of SECURITY_LEVEL, but it's here and changing the policy to say "ignored on client-side" is very very wrong.

**Comment by dlenski** [comment]
> For the use-case you mention, if the enterprises issues a certificate that is below the security posture of the employee's machine, then they should tell the employee how to change things. The client machine security posture must be part of the operating environment considered by the enterprise. In a BYOD world, employees might want stricter security, even if they have to "downgrade" it assert their company identity.

In practice, corporate VPNs are often poorly-maintained by people who don't really understand them, and don't care to think about security in sophisticated or systematic ways. They just need to check a box that says “we have client certificates” or “we have 128-bit encryption,” etc. I personally got involved with the OpenConnect project while working as a _consultant_, which meant that I was connecting to VPNs of clients where they had no control over my system, and I had no input into what kind of systems I might be asked to connect to.

An open-source VPN client like OpenConnect _can be_ a tool for achieving a secure connection when the server is configured sensibly, but often ends up simply as a tool for achieving a _compatible_ connection.

If I need to connect to a VPN that's running an old version of the Juniper Network Connect server software, with an ancient expired self-signed server certificate with an MD5 certificate, and which only supports 3DES as a cipher, and which might contain a backdoor due to the company's [alarming track record in this regard](https://en.wikipedia.org/wiki/Juniper_Networks#Security)… I shouldn't have any illusions about the “security” of the connection even though the provider of the VPN claims it exists for security purposes.

**Comment by dwmw2** [comment]
> I think saying that the application is consuming the certificate is using rather tortured logic to try to make a point.

No, it really isn't "tortured logic". As Dan's response hopefully makes clear, this is the painful reality. We need to do whatever nonsense is required of us by the VPN provider — including implementing compatibilty with pre-1.0 versions of the DTLS standard and other Faustian tortures.

A large proportion of that pain is around client certificates. These are provided to us in all manner of forms, including various file types, PKCS#11 tokens, TPM (v1.2 and v2.0) and other platform-specific stores. In an ideal world I'd take the string which the user gives me to describe the cert location, and just pass it over to the crypto library. In practice that doesn't *even* work for the various file types, all of which require explicit in detection in the app, and different procedures to load them.

And often the provided client cert comes without a full chain, the servers are configured without the intermediates too, and we have to jump through extra hoops to find our own issuer and explicitly add it to what's on the wire.

So no, saying that the cert is *provided* to us is absolutely not just "tortured logic". It is the merry hell of my existence as an application author.

And now to add insult to injury, we have a system-wide setting which is preventing my users from using the certificate they were issued, and which I cannot bypass just for *that* purpose, without downgrading the security level across the board.

I really don't understand why "apply this level *except* for using this client cert" is such a problematic nothing to ask for.

**Comment by richsalz** [comment]
So you have a system-wide setting that you cannot override for just your application (not sure that's true, but okay), and so therefore you want OpenSSL to carve out an except for your application and you don't see why it's problematic to ask for that.  Oh well.

**Comment by dwmw2** [comment]
I *can* override it completely for my application, just by unconditionally setting the security level to zero. It didn't seem like the right thing to do.

I don't consider my application to be a special case; it's just one of the broad class of SSL client applications. That I sometimes wish OpenSSL seemed to care about a little bit more.

**Comment by richsalz** [comment]
The problem is that SECURITY_LEVEL is too broad a brush.  Oh well.

**Comment by dlenski** [comment]
> … so therefore you want OpenSSL to carve out an except for your application and you don't see why it's problematic to ask for that.

I don't think anyone would describe examples like the ones David and I have cited as “unproblematic” :joy:.

They're simply a subset of systems using TLS in the real world (unfortunately a very _common_ subset), with the following characteristics:

1. System uses obsolete, known-vulnerable, and inefficient algorithms* or combines them in ill-advised ways.
2. The end user has little ability to do anything about the poor security of the system; the admins lack the time, expertise, money, or incentives to update it.
3. The end user _relies on_ being able to connect to the system. This is where well-designed, well-maintained client software focused on the end user can help a lot.

So yeah… these systems are _very_ problematic, but also very present in real-world applications. I believe that having the right API here can help us to connect to them with _as much security as possible_ (which is often, “not much security at all”) while maintaining functionality, and not reducing security for other users..

\* For a standout example, I formerly relied on access to a VPN whose server certificate expired in ~2008, only supports 3DES CBC with SHA1 MAC, _and_ only works with SSL v3.0 or [kinda-sorta-barely TLS 1.0](https://gitlab.com/gnutls/gnutls/-/issues/958#note_309267384).)

**Comment by richsalz** [comment]
I feel your pain, I really do.

**Comment by kroeckx** [comment]
So what is the problem with just using security level 0? It seems
that that is what you describe as what you need.

**Comment by dlenski** [comment]
> So what is the problem with just using security level 0? It seems that that is what you describe as what you need.

@kroeckx, agreed that “just use security level 0” solves the immediate problem.

What David is highlighting is that in the case of VPNs, there is a very common case of a connection which _would_ comply with SL1 _except_ that the client certificate is insecure. It would be great to be able to use OpenSSL's mechanism to verify the SL for “everything else.” In other words, some way to mark in the connection setup “yes we know that _this part which we're configuring right here_ is likely not to adhere to the desired SL, but can you still check that the rest of the setup complies?”

Without having looked at the implementation at all, it appears

**Comment by kroeckx** [comment]
Can that problem be sovled using SSL_set_security_callback?

**Comment by kaduk** [comment]
> We need to do whatever nonsense is required of us by the VPN provider — including implementing compatibilty with pre-1.0 versions of the DTLS standard and other Faustian tortures.

Of course!  We don't, however, necessarily need to say that it complies with security level N (for whatever value of N.)  Rich's assessment of security level "being a broad brush" seems accurate, and Dan's desire to have a tailored opt-out a reasonable desire (note: *not* the default behavior for a given security level).  IIUC Kurt's suggestion of `SSL_set_security_callback()` ought to work as such a tailored opt-out.

**Comment by dlenski** [comment]
> Can that problem be sovled using SSL_set_security_callback?

Might well be possible, yeah. Is there an explanation of what the op/bits/nid/other/ex arguments describe, or a code example?

```
void SSL_set_security_callback(SSL *s,
               int (*cb)(SSL *s, SSL_CTX *ctx, int op, int bits, int nid,
                                                       void *other, void *ex));
```

**Comment by kaduk** [comment]
You mean you can't just work from https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_security_level.html#APPLICATION-DEFINED-SECURITY-CALLBACKS :)

https://github.com/openssl/openssl/blob/6763f9c7e62aeba0d083be1608c88d85110976cb/apps/lib/s_cb.c#L1381 is a worked example that handles many possible cases to print debugging output (used by s_client/etc.)

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4  dev, barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
